### ISSUE TITLE
 S3TransferManager creation fails on Apple's M1

### ISSUE THREAD

**Issue by AdamBien** [aws/aws-sdk-java-v2#2942]
### Describe the bug

Instantiation of S3TransferManager fails

### Expected behavior

S3TransferManager is instantiated, file is uploaded.

### Current behavior

The instantiation of fails. 

var presigner = S3Presigner.builder().build();
var credentialsProvider = StaticCredentialsProvider.create(credentials);
var configuration = S3ClientConfiguration
        .builder()
        .region(Region.EU_CENTRAL_1)
        .credentialsProvider(credentialsProvider)
    .build();

This file below fails:
var transferManager = S3TransferManager.builder().s3ClientConfiguration(configuration).build();

with:

Unable to unpack AWS CRT lib: java.io.IOException: Unable to open library in jar for AWS CRT: /osx/armv8/libaws-crt-jni.dylib
java.io.IOException: Unable to open library in jar for AWS CRT: /osx/armv8/libaws-crt-jni.dylib
        at software.amazon.awssdk.crt.CRT.extractAndLoadLibrary(CRT.java:158)
        at software.amazon.awssdk.crt.CRT.loadLibraryFromJar(CRT.java:223)
        at software.amazon.awssdk.crt.CRT.<clinit>(CRT.java:33)
        at software.amazon.awssdk.crt.CrtResource.<clinit>(CrtResource.java:100)
        at software.amazon.awssdk.transfer.s3.internal.S3NativeClientConfiguration.<init>(S3NativeClientConfiguration.java:58)
        at software.amazon.awssdk.transfer.s3.internal.S3NativeClientConfiguration$Builder.build(S3NativeClientConfiguration.java:194)
        at software.amazon.awssdk.transfer.s3.internal.DefaultS3CrtAsyncClient.<init>(DefaultS3CrtAsyncClient.java:54)
        at software.amazon.awssdk.transfer.s3.internal.DefaultS3CrtAsyncClient$DefaultS3CrtClientBuilder.build(DefaultS3CrtAsyncClient.java:224)
        at software.amazon.awssdk.transfer.s3.internal.DefaultS3TransferManager.initializeS3CrtClient(DefaultS3TransferManager.java:95)
        at software.amazon.awssdk.transfer.s3.internal.DefaultS3TransferManager.<init>(DefaultS3TransferManager.java:60)
        at software.amazon.awssdk.transfer.s3.internal.DefaultS3TransferManager$DefaultBuilder.build(DefaultS3TransferManager.java:310)

### Steps to Reproduce

import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider;
import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.s3.model.PutObjectRequest;
import software.amazon.awssdk.services.s3.presigner.S3Presigner;
import software.amazon.awssdk.services.s3.presigner.model.PutObjectPresignRequest;
import software.amazon.awssdk.transfer.s3.S3ClientConfiguration;
import software.amazon.awssdk.transfer.s3.S3TransferManager;

var presigner = S3Presigner.builder().build();
var credentialsProvider = StaticCredentialsProvider.create(credentials);
var configuration = S3ClientConfiguration
        .builder()
        .region(Region.EU_CENTRAL_1)
        .credentialsProvider(credentialsProvider)
    .build();

var transferManager = S3TransferManager.builder().s3ClientConfiguration(configuration).build();

### Possible Solution

S3Client was used for upload

### Context

I tried the new transfer manager to upload a file to S3 bucket and generate a presigned URL afterwards. The transfer manager relies on CRT which does not ship with the maven dependency and causes the error above.

### AWS Java SDK version used

2.17.102 / 2.17.102-PREVIEW

### JDK version used

Tested with java version "17.0.1" 2021-10-19 LTS and OpenJDK Runtime Environment Corretto-17.0.1.12.1 (build 17.0.1+12-LTS)

### Operating System and version

macOS Monterey 12.0.1 (Apple M1)

**Comment by zoewangg** [comment]
Hi @AdamBien thank you for reporting the issue. We will investigate it

**Comment by RobertoUa** [comment]
any news?

**Comment by zoewangg** [comment]
Sorry, no update at the moment. This issue is still on our radar, and we will post here once we have more information.

**Comment by mschwartau** [comment]
Seems to be a problem in the aws-crt jar because it doesn't contain a lib for m1 macs (it contains one for x86 macs, linux and so on).

There is already a bug for this in the crt lib: https://github.com/awslabs/aws-crt-java/issues/371. The workaround described in the issue worked for me:
- Clone the project
- build the [aws-crt-java](https://github.com/awslabs/aws-crt-java) project locally (see readme for details)
- Now use the jar file from the local maven repo, it contains the `osx/armv8/libaws-crt-jni.dylib` which is missing in the official build
- Now the TransferManager can be created.

So the issue https://github.com/awslabs/aws-crt-java/issues/371 just needs to be fixed - and after that the new version of aws-crt has to be used.

**Comment by rhanak** [comment]
We are also having this issue using the `AwsCrtAsyncHttpClient` on M1 mac machines. Fails with the following but works when we go with the workaround above to build aws-crt locally. Would be great if this could be fixed in CI and maven package updated!
```
Unable to unpack AWS CRT lib: java.io.IOException: Unable to open library in jar for AWS CRT: /osx/armv8/libaws-crt-jni.dylib
```

**Comment by adamalfar** [comment]
This is still a problem (mac m1 Max Monterey 12.3.1).  I get the same error as above (Unable to unpack AWS CRT lib: java.io.IOException: Unable to open library in jar for AWS CRT: /osx/armv8/libaws-crt-jni.dylib) when I run Transfer Manager example (using  <version>2.17.197-SNAPSHOT-PREVIEW</version> )

**Comment by zoewangg** [comment]
Hi @adamalfar, thanks for reporting this issue. Just to double check, the snapshot version you were testing against contains this change https://github.com/aws/aws-sdk-java-v2/pull/3209/commits/a3438b0cb833eb4e52a170d7bcda6cf4196d4845, correct?

**Comment by adamalfar** [comment]
Yes @zoewangg.  Thanks!

**Comment by adamalfar** [comment]
Here is more detailed Exception log:
Unable to unpack AWS CRT lib: java.io.IOException: Unable to open library in jar for AWS CRT: /osx/armv8/libaws-crt-jni.dylib
java.io.IOException: Unable to open library in jar for AWS CRT: /osx/armv8/libaws-crt-jni.dylib
	at software.amazon.awssdk.crt.CRT.extractAndLoadLibrary(CRT.java:161)
	at software.amazon.awssdk.crt.CRT.loadLibraryFromJar(CRT.java:226)
	at software.amazon.awssdk.crt.CRT.<clinit>(CRT.java:33)
	at software.amazon.awssdk.crt.CrtResource.<clinit>(CrtResource.java:104)
	at software.amazon.awssdk.transfer.s3.internal.S3NativeClientConfiguration.<init>(S3NativeClientConfiguration.java:60)
	at software.amazon.awssdk.transfer.s3.internal.S3NativeClientConfiguration$Builder.build(S3NativeClientConfiguration.java:208)
	at software.amazon.awssdk.transfer.s3.internal.S3CrtAsyncHttpClient.<init>(S3CrtAsyncHttpClient.java:61)
	at software.amazon.awssdk.transfer.s3.internal.S3CrtAsyncHttpClient.<init>(S3CrtAsyncHttpClient.java:47)
	at software.amazon.awssdk.transfer.s3.internal.S3CrtAsyncHttpClient$Builder.build(S3CrtAsyncHttpClient.java:242)
	at software.amazon.awssdk.transfer.s3.internal.DefaultS3CrtAsyncClient.<init>(DefaultS3CrtAsyncClient.java:66)
	at software.amazon.awssdk.transfer.s3.internal.DefaultS3CrtAsyncClient.<init>(DefaultS3CrtAsyncClient.java:54)
	at software.amazon.awssdk.transfer.s3.internal.DefaultS3CrtAsyncClient$DefaultS3CrtClientBuilder.build(DefaultS3CrtAsyncClient.java:223)
	at software.amazon.awssdk.transfer.s3.internal.DefaultS3TransferManager.initializeS3CrtClient(DefaultS3TransferManager.java:111)
	at software.amazon.awssdk.transfer.s3.internal.DefaultS3TransferManager.<init>(DefaultS3TransferManager.java:75)
	at software.amazon.awssdk.transfer.s3.internal.DefaultS3TransferManager$DefaultBuilder.build(DefaultS3TransferManager.java:425)
	at S3TransferManagerTest.main(S3TransferManagerTest.java:24)
Exception in thread "main" java.lang.ExceptionInInitializerError
	at software.amazon.awssdk.crt.CrtResource.<clinit>(CrtResource.java:104)
	at software.amazon.awssdk.transfer.s3.internal.S3NativeClientConfiguration.<init>(S3NativeClientConfiguration.java:60)
	at software.amazon.awssdk.transfer.s3.internal.S3NativeClientConfiguration$Builder.build(S3NativeClientConfiguration.java:208)
	at software.amazon.awssdk.transfer.s3.internal.S3CrtAsyncHttpClient.<init>(S3CrtAsyncHttpClient.java:61)
	at software.amazon.awssdk.transfer.s3.internal.S3CrtAsyncHttpClient.<init>(S3CrtAsyncHttpClient.java:47)
	at software.amazon.awssdk.transfer.s3.internal.S3CrtAsyncHttpClient$Builder.build(S3CrtAsyncHttpClient.java:242)
	at software.amazon.awssdk.transfer.s3.internal.DefaultS3CrtAsyncClient.<init>(DefaultS3CrtAsyncClient.java:66)
	at software.amazon.awssdk.transfer.s3.internal.DefaultS3CrtAsyncClient.<init>(DefaultS3CrtAsyncClient.java:54)
	at software.amazon.awssdk.transfer.s3.internal.DefaultS3CrtAsyncClient$DefaultS3CrtClientBuilder.build(DefaultS3CrtAsyncClient.java:223)
	at software.amazon.awssdk.transfer.s3.internal.DefaultS3TransferManager.initializeS3CrtClient(DefaultS3TransferManager.java:111)
	at software.amazon.awssdk.transfer.s3.internal.DefaultS3TransferManager.<init>(DefaultS3TransferManager.java:75)
	at software.amazon.awssdk.transfer.s3.internal.DefaultS3TransferManager$DefaultBuilder.build(DefaultS3TransferManager.java:425)
	at S3TransferManagerTest.main(S3TransferManagerTest.java:24)
Caused by: software.amazon.awssdk.crt.CrtRuntimeException: software.amazon.awssdk.crt.CrtRuntimeException: Unable to unpack AWS CRT library UNKNOWN(-1) UNKNOWN(-1)
	at software.amazon.awssdk.crt.CRT.loadLibraryFromJar(CRT.java:236)
	at software.amazon.awssdk.crt.CRT.<clinit>(CRT.java:33)
	... 13 more

**Comment by zoewangg** [comment]
Thanks, we are not able to reproduce the issue unfortunately. Could you try with the released version `2.17.198`? Could you also confirm that the CRT version resolved is 0.16.10? You can verify it by running this command `mvn dependency:tree -Dincludes=software.amazon.awssdk.crt:aws-crt`

**Comment by adamalfar** [comment]
Tried with 2.17.198 rel
Here is the dep tree for CRT

[INFO] --- maven-dependency-plugin:3.1.1:tree (default-cli) @ s3 ---
[INFO] software.amazon.awssdk:s3:jar:2.17.198-SNAPSHOT
[INFO] \- software.amazon.awssdk:auth-crt:jar:2.17.198-SNAPSHOT:test
[INFO]    \- software.amazon.awssdk.crt:aws-crt:jar:0.16.10:test

**Comment by adamalfar** [comment]
Ok this works now.  Amazing power of "restarting" the machine.  Thanks to all!

**Comment by zoewangg** [comment]
@adamalfar glad to hear it works now! Thank you again for verifying it! 

Hey all, M1 support has been released in `2.17.198`. I'm going to close the issue, but feel free to open new GH issues if you have questions.
